src/test/java/com/google/firebase/auth/FirebaseAuthIT.java

```diff
@@ -140,37 +141,92 @@ public void testDeleteNonExistingUser() throws Exception {
       }
       assertTrue(providers.contains("password"));
       assertTrue(providers.contains("phone"));
 
-      checkRecreate(randomUser.uid);
+      checkRecreate(user.getUid());
+    } finally {
+      auth.deleteUserAsync(randomUser.uid).get();
+    }
+  }
+
+  @Test
+  public void testLookupUserByPhone() throws Exception {
+    RandomUser randomUser1 = RandomUser.create();
+    RandomUser randomUser2 = RandomUser.create();
+    RandomUser randomUser3 = RandomUser.create();
+    try {
+      UserRecord user1 = createUser(
+          randomUser1.uid, /* phoneNumber= */ null, randomUser1.email);
+      UserRecord user2 = createUser(
+          randomUser2.uid, randomUser2.phone, randomUser2.email);
+      UserImportResult user3 = importUser(
+          randomUser3.uid, randomUser3.phone, randomUser3.email,
+          "google.com", randomUser3.uid + "_google.com");
+
+      UserRecord lookedUpRecord = auth.getUserByPhoneNumberAsync(randomUser2.phone).get();
+      assertEquals(randomUser2.uid, lookedUpRecord.getUid());
+
+      lookedUpRecord = auth.getUserByPhoneNumberAsync(randomUser3.phone).get();
+      assertEquals(randomUser3.uid, lookedUpRecord.getUid());
+    } finally {
+      auth.deleteUserAsync(randomUser1.uid).get();
+      auth.deleteUserAsync(randomUser2.uid).get();
+      auth.deleteUserAsync(randomUser3.uid).get();
+    }
+  }
+
+  @Test
+  public void testLookupUserByFederatedId() throws Exception {
+    RandomUser randomUser1 = RandomUser.create();
+    RandomUser randomUser2 = RandomUser.create();
+    RandomUser randomUser3 = RandomUser.create();
+    try {
+      UserRecord user1 = createUser(
+          randomUser1.uid, /* phoneNumber= */ null, randomUser1.email);
+      UserRecord user2 = createUser(
+          randomUser2.uid, randomUser2.phone, randomUser2.email);
+      UserImportResult user3 = importUser(
+          randomUser3.uid, randomUser3.phone, randomUser3.email,
+          "google.com", randomUser3.uid + "_google.com");
+
+      UserRecord lookedUpRecord = auth.getUserByFederatedIdAsync(
+          randomUser3.uid + "_google.com", "google.com").get();
+      assertEquals(randomUser3.uid, lookedUpRecord.getUid());
+      assertEquals(2, lookedUpRecord.getProviderData().length);
+      List<String> providers = new ArrayList<>();
+      for (UserInfo provider : lookedUpRecord.getProviderData()) {
+        providers.add(provider.getProviderId());
+      }
+      assertTrue(providers.contains("phone"));
+      assertTrue(providers.contains("google.com"));
+
+      try {
+        // Verify that lookup by federated identifier does not accept "phone".
+        lookedUpRecord = auth.getUserByFederatedIdAsync(
+            randomUser3.phone, "phone").get();
+        fail("No error thrown for non-federated provider");
+      } catch (IllegalArgumentException expected) {
+      }
     } finally {
-      auth.deleteUserAsync(userRecord.getUid()).get();
+      auth.deleteUserAsync(randomUser1.uid).get();
```

Comment:
We should use bulk delete for this when it becomes available. Add a TODO so we don't forget.